### ISSUE TITLE
How to prevent empty clusters?

### ISSUE THREAD

**Issue by LiXirong** [facebookresearch/deepcluster#12]
Hi~

I use sklearn.cluster.Kmeans, PCA, and pytorch to do the same job on my own dataset.

And I find that when I clustering for each epoch, there are many empty clusters as said in this paper 3.3.

But I don't find the solution on this repository. I want to know what can I do exactly to prevent empty clusters? 

Thanks!

**Comment by mathildecaron31** [comment]
Hi,
Thanks for your interest!
What you can do is sampling the batches based on a uniform distribution over the clusters (see [here](https://github.com/facebookresearch/deepcluster/blob/f5995e954842054d88aa9fcc9ff7ba2db7eafc9e/util.py#L49.). You may also split one cluster in two whenever you get an empty cluster or add an entropic term in your k-means loss.

**Comment by yassouali** [comment]
Hi, 

In the provided implementation, is the avoidance of empty clusters by reassigning the empty clusters is done automatically by faiss clustering?

Thanks

**Comment by mathildecaron31** [comment]
Hi, 
Thanks for your interest and sorry for the delay of my reply.
Nothing prevents empty clusters from occurring in my implementation. This is not done automatically by faiss.
The significance of each class labels changes from a clustering assignment to another, that's why we reset the last fully connected layers. It is an expected behavior.
Please re-open the issue if you have further questions.

**Comment by skr998** [comment]
Hi,@yassouali
I have a same problem about the reassignment of the labels between each two epochs. which is the new paper you said, Thank you.
